sdk/core/Azure.Core/tests/TestFramework/TestEnvironment.cs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Core.Testing
+{
+    /// <summary>
+    ///   Represents the ambient environment in which the test suite is
+    ///   being run.
+    /// </summary>
+    public partial class TestEnvironment
+    {
+        private TokenCredential _credential;
+        private readonly string _prefix;
+
+        public TestEnvironment(string serviceName)
+        {
+            _prefix = serviceName.ToUpperInvariant() + "_";
+        }
+
+        /// <summary>
+        ///   The name of the Azure subscription containing the resource group to be used for Live tests. Recorded.
+        /// </summary>
+        public string SubscriptionId => GetRecordedVariable("SUBSCRIPTION_ID");
+
+        /// <summary>
+        ///   The name of the Azure resource group to be used for Live tests. Recorded.
+        /// </summary>
+        public string ResourceGroup => GetRecordedVariable("RESOURCE_GROUP");
+
+        /// <summary>
+        ///   The location of the Azure resource group to be used for Live tests (e.g. westus2). Recorded.
+        /// </summary>
+        public string Location => GetRecordedVariable("LOCATION");
+
+        /// <summary>
+        ///   The environment of the Azure resource group to be used for Live tests (e.g. AzureCloud). Recorded.
+        /// </summary>
+        public string AzureEnvironment => GetRecordedVariable("ENVIRONMENT");
+
+        /// <summary>
+        ///   The name of the Azure Active Directory tenant that holds the service principal to use during Live tests. Recorded.
+        /// </summary>
+        public string TenantId => GetRecordedVariable("TENANT_ID");
+
+        /// <summary>
+        ///   The client id of the Azure Active Directory service principal to use during Live tests. Recorded.
+        /// </summary>
+        public string ClientId => GetRecordedVariable("CLIENT_ID");
+
+        /// <summary>
+        ///   The client secret of the Azure Active Directory service principal to use during Live tests. Not recorded.
+        /// </summary>
+        public string ClientSecret => GetVariable("CLIENT_SECRET");
+
+
+        public TokenCredential Credential
+        {
+            get
+            {
+                if (_credential != null)
```

Comment:
Should this be thread-safe, or we expect that no test will access this property from several threads concurrently?

Reply:
Even if they access it concurrently they all get equivalent instances and some of them would get GCed.